type-stable inner loop for sqrtm

felixrehren opened the pull request:
As suggested by Ralph_Smith on [discourse](https://discourse.julialang.org/t/review-schur-pade-matrix-powers-speedup/1650/6)

On my machine: speedup x15

felixrehren committed:
type-stable inner loop for sqrtm

As suggested by Ralph_Smith on [discourse](https://discourse.julialang.org/t/review-schur-pade-matrix-powers-speedup/1650/6)

On my machine: speedup x15

vchuravy commented:
Do we have a benchmark for this in BaseBenchmarks?

felixrehren commented:
I can't find a benchmark. What would that look like?

    function b_20214(n=100,m=127)
        t = 0
        for i in 1:n
            A = randn(m,m)
            B = UpperTriangular(A'*A)
            t += @timeit sqrtm(B)
        end
        t
    end

andreasnoack commented:
Take a look at https://github.com/JuliaCI/BaseBenchmarks.jl#contributing and https://github.com/JuliaCI/BaseBenchmarks.jl/blob/master/src/linalg/LinAlgBenchmarks.jl. The package will handle the timing. You should just provide the function.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1888,10 +1898,7 @@ function sqrtm{T}(A::UpperTriangular{T})
     for j = 1:n
```
It seems like we should just have `sqrtm(A)` call a `sqrtm{realmatrix}(A,::Val{realmatrix})` function with either `Val{true}` or `Val{false}` so that every operation on `R` is type-stable, and you don't need `floop`.

stevengj replied:
You can put `@inbounds` in front of this loop.

felixrehren replied:
Done both :+1:

pabloferz replied:
You can just put the `@inbounds` in front of the `for ... end` group no need of the extra `begin ... end` here. That is,

```julia
@inbounds for j = 1:n
    # loop body
end
```

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1900,14 +1907,10 @@ end
 function sqrtm{T}(A::UnitUpperTriangular{T})
```
It seems like we could replace this with
```jl
sqrtm{T}(A::UnitUpperTriangular{T}) = sqrtm(A, Val{true})
```
in terms of the abovementioned method.  Given a `Val` method that is type-stable, the only remaining reason to duplicate the code here seems to be to save the `sqrt` call for the diagonal elements, but I doubt this is worth the trouble since there are only O(n) such calls.

felixrehren replied:
Will look at this tomorrow

felixrehren replied:
Since `UnitUpperTriangular{T}` is not a subtype of `UpperTriangular{T}` (??), I think it would be

    sqrtm(A::UnitUpperTriangular) = UnitUpperTriangular(sqrtm(UpperTriangular(A),Val{true}))

For some reason this is *faster* than the present method, which confuses me. This version involves two additional conversions and some unnecessary arithmetic operations in the loop (of order O(n) as you mention)! Are some ops on `UnitUpperTriangular` slower than the same op on `UpperTriangular`?

Will make this change later unless someone can enlighten me

stevengj replied:
(Julia doesn't support inheritance of concrete types.)

Indexing expressions `A[i,j]` are slower for `UnitUpperTriangular` than for `UpperTriangular`, and both are slower than for `Array`, because it has to check if `i <= j` and (in the unit case) whether `i == j`.

Since by construction you only access the upper triangle, I would do:
```jl
B = A.data
```
at the beginning of the function (for both the `UnitUpperTriangular` and `UpperTriangular` methods), and then use `B[i,j]` rather than `A[i,j]`.

stevengj replied:
(There might be other methods operating on triangular matrix types that might benefit from a similar change.)

andreasnoack replied:
I think most of the methods in `Base.LinAlg` upwrap the type before indexing. Haven't checked, though.

felixrehren replied:
I'm lost. I see a small improvement, ~10%, doing this for the `UnitUpperTriangular` method, and it now has the same performance as converting to `UpperTriangular` and back. I see no improvement at all for `UpperTriangular` methods. But all the array access is already wrapped in `@inbounds` anyway, so is there really still `i<=j` testing going on? And `@time`ing `A[i,j]` versus `B[i,j]` shows no difference, no matter the type of `A` ...

stevengj replied:
The `@inbounds` means that it never checks where `i` and `j` are in `1:n`.  It still checks whether `i ≤ j`, since if `i > j` then it just returns zero rather than looking at `A.data`.

Timing such a small operation is tricky; I would normally recommend BenchmarkTools instead of `@time`:
```julia
julia> using BenchmarkTools

julia> A = rand(10,10); T = UpperTriangular(A);

julia> @btime $A[3,5]; @btime $T[3,5];
  1.377 ns (0 allocations: 0 bytes)
  1.650 ns (0 allocations: 0 bytes)
```

stevengj replied:
(It may be that the effect on the `UpperTriangular` case is too small to measure, however.  Or maybe even LLVM is smart enough to figure out that `i ≤ j` is always true.  But it is probably still a good idea to unwrap the type.)

felixrehren replied:
After improving the type stability, the specialised implementation for `UnitUpperTriangular` is faster than converting.
Now, on 0.5.0 using `@benchmark`ing, I see a *slowdown* when unwrapping the type first in `sqrtm`. Relatedly, on my machine (juliabox.com) there appears to be a slowdown for the atomic operations:

    @benchmark $A[1,2]
      median time:      2.650 ns
    @benchmark $T[1,2]
      median time:      2.223 ns

(what does the `$` do here?)
This persists when wrapping these calls in a function. All in all I am hesitant to make the unwrapping change for `sqrtm` (but not against)

stevengj replied:
How did you unwrap the type?  I hope you didn't do `A = A.data`, which hurts type stability.

I can't reproduce your benchmark results.  Maybe the benchmarks aren't reliable for such a short operation, and one needs to put a bunch of indexing operations in a loop?  (Probably you also want to time them with `@inbounds`).   The `$` splices the value of the variable directly into the expression, which gets rid of the penalty for benchmarking with a global variable (it means the compiler does not have to do runtime type inference to find the type of `A`).

stevengj replied:
For example:
```julia
julia> function sumupper(A)
           s = zero(eltype(A))
           for j = 1:size(A,2), i = 1:j
               s += A[i,j]
           end
           return s
       end
sumupper (generic function with 1 method)

julia> A = rand(100,100); T = UpperTriangular(A);

julia> using BenchmarkTools

julia> sumupper(A) == sumupper(T)
true

julia> @btime sumupper($A); @btime sumupper($T);
  4.122 μs (0 allocations: 0 bytes)
  5.102 μs (0 allocations: 0 bytes)
```

felixrehren committed:
dispatch sqrtm on real-or-not bool

As suggested by @stevengj

felixrehren commented:
@stevengj great idea :+1: In my testing, I found that `floop` still offered a performance improvement (that I couldn't replicate by "manually inlining" `floop`). Not sure why that would be, but for the time being I left the use of `floop` as is. Happy to modify further.

@andreasnoack I'm preparing a PR for BaseBenchmarks :+1:

tkelman commented on base/linalg/triangular.jl:
```diff
@@ -1889,14 +1888,18 @@ function sqrtm{T}(A::UpperTriangular{T})
             end
         end
     end
+    sqrtm(A::UpperTriangular,Val{realmatrix})
```
the `::UpperTriangular` type assertion isn't usually used on call sites

felixrehren committed:
remove call site type assertion

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1879,19 +1888,20 @@ function sqrtm{T}(A::UpperTriangular{T})
     if realmatrix
         TT = typeof(sqrt(zero(T)))
     else
         TT = typeof(sqrt(complex(-one(T))))
```
just use `sqrt(complex(zero(T)))` here, since at some point `one` may return a dimensionless value if `T` is dimensionful.

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1867,7 +1867,16 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
 
-function sqrtm{T}(A::UpperTriangular{T})
+function floop(x,R,i::Int,j::Int)
+    r = x
```
This is not type stable if `A[i,j]` and `R[i,j]` are not the same type.   A simple fix is `r = x + zero(eltype(R))`

stevengj commented:
@felixrehren, have you tried `@code_warntype` to see if there are any type instabilities in `sqrtm(A, Val{...})`?

felixrehren committed:
more type stability

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1894,15 +1884,20 @@ function sqrtm{T,realmatrix}(A::UpperTriangular{T},::Type{Val{realmatrix}})
+        for j = 1:n
+            R[j,j] = realmatrix ? sqrt(A[j,j]) : sqrt(complex(A[j,j]))
+            for i = j-1:-1:1
+                r = A[i,j] + zero(TT)
```
Even simpler:
```jl
r::TT = A[i,j]
```
Though, like a lot of our generic linear-algebra code, this isn't right for dimensionful quantities, because `r` has the units of `A` whereas `R` has the units of `sqrt(A)`.  If we want to get that right, it should really be something like:
```
TA = realmatrix ? float(T) : Complex{float(T)}
```
and then use `r::TA`.

felixrehren replied:
To be clear, do you suggest replacing

    TT = realmatrix ? typeof(sqrt(zero(T))) : typeof(sqrt(complex(zero(T))))
with

    TA = realmatrix ? float(T) : Complex{float(T)}
? (And then later using `r::TA` or `r::TT` in either case) Sounds good to me, much clearer :+1:

stevengj replied:
`TA` would be supplemental to `TT`, because `A` and `R` have different units.

andreasnoack replied:
Instead of introducing a new type variable then you can simply do
```
r = A[i,j] - zero(TT)*zero(TT)
```
which matches the operation in the loop and therefore correct. We should avoid explicit `float` calls if possible since it requires that float is implemented correctly for new types (which might not be the case).

felixrehren replied:
I hope this is addressed -- I guess we should write some tests? (for a different PR ...)

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1894,15 +1884,20 @@ function sqrtm{T,realmatrix}(A::UpperTriangular{T},::Type{Val{realmatrix}})
     end
-    n = checksquare(A)
+    n = Base.LinAlg.checksquare(A)
     R = zeros(TT, n, n)
```
My inclination would be to use `R = Matrix{TT}(n, n)` here.  Then get rid of the `r == 0` check below.  `sqrtm` is unlikely to return a sparse result, so I don't see the point of pre-initializing the array to zero.

stevengj replied:
Oh, nevermind, I forgot that this is for the upper-triangular case.  Then I guess initializing it to zero makes sense.

felixrehren commented:
@stevengj Thanks for the comments, I am also learning a lot!
I used @code_warntype and, with the suggestions you made that are already implemented, there are no more warnings or unions in the output of `sqrtm(A,Val{realmatrix})`

I think the AppVeyor failure is unrelated?

    Error in testset subarray:
    Error During Test
      Test threw an exception of type ErrorException
      Expression: X[1,1:end - 2] == @views(X[1,1:end - 2])
      Partial linear indexing is deprecated. Use `reshape(A, Val{2})` to make the dimensionality of the array match the number of indices.

stevengj commented:
The "Partial linear indexing" thing was fixed by #20242

felixrehren committed:
casting changes

replace `TT` by `t` for the type of the sqrt of a variable of type `T`
introduce `tt` as the type of the square of a variable of type `t`
N.B. `tt` is not always the same as `T`, it could be `Complex{T}`
In the `UnitUpperTriangular` case, some of the complexity should fall away; TODO?

tkelman commented on base/linalg/triangular.jl:
```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = r/one)
```
`one` won't be equivalent to the former `(R[i,i] + R[j,j])`, will it?

felixrehren replied:
You're right, thanks!

tkelman replied:
why is it equivalent to `2*R[1,1]` ?

felixrehren replied:
Because this is for the `UnitUpperTriangular` case only, `R[i,i] = R[1,1]`

tkelman replied:
Ah right. Missed the `UnitUpperTriangular` since it was on the other side of a long conversation.

felixrehren committed:
one -> two

correct previous change, thx tkelman

felixrehren committed:
div 2 -> mult 1/2

felixrehren committed:
typo

ViralBShah commented:
I know this is not a blocker for 0.6, but would be nice to get it in as we plan for an alpha release.

tkelman commented:
Please don't put "nice to have" on the milestone.

felixrehren commented:
Think this is it -- @stevengj, @andreasnoack ?

tkelman commented on base/linalg/triangular.jl:
```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = half*r)
```
the old version of this function might not have worked correctly in this case either, but would this do the right thing if the element type was not commutative under multiplication?

stevengj replied:
See my comment above for the general case.  However, even for non-commutative algebras the case here is special because `R[i,i]` is the identity, which always commutes with everything.  So, the problem reduces to solving `R[i,j] + R[i,j] = r`.  *If* the elements form a field over the reals, then the solution `R[i,j] = r/2` is correct.

The case here is also wrong if e.g. the elements are matrices, in which case `1/(2*R[1,1])` will throw an error, although if you do `half = inv(2*R[1,1])` it will work.   However, it would be even better to just to `R[i,j] = r/2` in that case, since it would avoid the matrix multiplication.

However, it is still wrong if the elements are some other number field where you can't do `/2`. The correct, general solution should be
```julia
half=inv(R[1,1] + R[1,1])
```

stevengj commented:
(Note that this PR can go in after the feature freeze since it is just a performance optimization.)

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+                @simd for k = i+1:j-1
+                    r -= R[i,k]*R[k,j]
+                end
+                r==0 || (R[i,j] = r / (R[i,i] + R[j,j]))
```
Per @tkelman's comment below, for non-commutative number types (e.g. quaternions), this is not right.

I haven't gone through the algebra very carefully, but I think the final `R[i,j]` needs to solve `R[i,i]*R[i,j] + R[i,j]*R[j,j] == r`, i.e. a [Sylvester equation](https://en.wikipedia.org/wiki/Sylvester_equation).  So, the right thing to do seems like:
```julia
R[i,j] = sylvester(R[i,i], R[j,j], -r)
```
and then add a method `sylvester(a::Union{Real,Complex},b::Union{Real,Complex},c::Union{Real,Complex}) = -c / (a + b)` for the commutative `Number` cases.  New `Number` types will have then to provide their own `sylvester` method if they want `sqrtm` to work.
```

felixrehren committed:
algebraicness, unwrapping

stevengj commented on base/linalg/triangular.jl:
```diff
@@ -1867,48 +1867,53 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
+    sqrtm(A,Val{realmatrix})
+end
+# solve the sylvester equation a*x + x*b + c for x when a,b,x are commutative numbers. PR#20214
+sylvester(a::Union{Real,Complex},b::Union{Real,Complex},c::Union{Real,Complex}) = -c / (a + b)
```
Probably this `sylvester` definition should go into complex.jl or similar?  And then you'll need to make sure that the LinAlg module imports `Base.sylvester` so that it extends the base definition.

stevengj replied:
I would parenthesize `(-c)`.  That way if you call `sylvester(a, b, -r)`, the compiler will have a good shot at noticing that the two negations cancel.

stevengj replied:
Hmm, it looks like LLVM is smart enough to eliminate the double negation either way:
```jl
julia> syl(a,b,c) = -c / (a + b)
syl (generic function with 1 method)

julia> foo(a,b,c) = syl(a,b,-c)
foo (generic function with 1 method)

julia> @code_llvm foo(1.0,2.0,3.0)

define double @julia_foo_65538(double, double, double) #0 !dbg !5 {
top:
  %3 = fadd double %0, %1
  %4 = fdiv double %2, %3
  ret double %4
}
```

stevengj commented:
LGTM except that the definition of `sylvester` should go elsewhere.  Maybe in `linalg/dense.jl` with the other `sylvester` methods for now.

stevengj commented:
Also, you don't need to put `PR #....` comments on lines... that's what `git blame` is for.

felixrehren commented:
@stevengj Ok, will do. Should I keep any of the comments or clean it all up?

stevengj commented:
You should have a comment on code that does something in a non-obvious way, to make sure no one changes it without realizing, but your comments don't have to mention the PR.

felixrehren committed:
sylvester for numbers [ci skip]

moved from triangular.jl

felixrehren committed:
move sylvester

felixrehren commented:
Think all comments so far are addressed, let me know if it could use further improvements

stevengj commented:
LGTM.  Is the overall speedup still 15x?

felixrehren commented:
With all the nice tweaks, I see x20 [in this benchmarking notebook](https://github.com/felixrehren/julia/blob/fr-sqrtbench/sqrtm-bench.ipynb) on Julia 0.5
